### ISSUE TITLE
Request: Fade in/out browser window

### ISSUE THREAD

**Issue by tommoor** [atom/electron#2407]
On OSX it is common for secondary windows to fade in and out.

I was wondering if this is something that others would like to see support for? It could be implemented as a new API or perhaps an additional parameter to show/hide methods.

**Comment by iamdriz** [comment]
@tommoor Can you not use JavaScript / CSS Animations to fade the contents?

**Comment by tommoor** [comment]
@iamdriz this would result in a quite different effect - definitely something that belongs at an OS level :smile:

**Comment by iamdriz** [comment]
Can you show a screenshot of what you mean by native fade out of secondary windows? I'm a Mac user and not sure what you mean. Thanks.

**Comment by tommoor** [comment]
Best I can do in GIF form...

![2015-08-06 08_25_19](https://cloud.githubusercontent.com/assets/380914/9115533/8dedc5e6-3c14-11e5-9228-0e223b56d43e.gif)

**Comment by iamdriz** [comment]
@tommoor You can't create that type of window in Electron anyway can you? All windows all the standard windows. To create that style, you would need to build it yourself in HTML/CSS wouldn't you?

In any case you could animate this by listening to the focus and blur events and then animating the window contents. And then create your own animations, etc.

**Comment by tommoor** [comment]
@iamdriz you can make something that looks like a crappy copy sure, but then that means you have to use the transparent browser-window option which comes with *many* drawbacks. 

HTML is great for all the stuff inside the bounds of the window - but not for outside the window, once you get into drawing fake shadows, fake borders, fake rounding edges, fake window controls it's a bad path to go down.

This new option could also be used on framed windows!

**Comment by iamdriz** [comment]
@tommoor Agreed that the fade effect should be native if you are trying to replicate an OS specific effect. But I suggested the CSS/JS route because this window type isn't supported in Electron. I don't think the fade should be an option though, rather the OS chooses to animate it, if the type of window you have chosen animates native or not if it doesn't.

**Comment by paulcbetts** [comment]
> @iamdriz you can make something that looks like a crappy copy sure, but then that means you have to use the transparent browser-window option which comes with many drawbacks in Electron.

Actually if you're only doing this on OS X, you can probably get away with it, it's by far the best platform for transparent browser windows

**Comment by jorangreef** [comment]
| definitely something that belongs at an OS level :smile:

I would also really appreciate a native fadeIn and fadeOut method on BrowserWindows, where the OS supports it.

We recently switched to using a BrowserWindow instead of a Tray menu because the current implementation of the Tray menu prevents item labels from being updated while the menu is visible, and resetting the menu causes the menu to close.

**Comment by zcbenz** [comment]
@bengotow I prefer a Node module if it is JavaScript-based, but it would be great if we can a get a [native implementation](https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/AnimationGuide/Articles/ViewAnimations.html).

**Comment by bengotow** [comment]
Yeah—I have a hunch that performance would be much better on Mac OS X if the animation was queued directly in CoreAnimation, but we're getting ~25ish FPS with our current "setBounds over and over again in JavaScript" approach so that may be easiest. Definitely need to run some experiments.

**Comment by davej** [comment]
@bengotow: Just wondering, did you check to see if there is a performance hit incurred by doing this using the `remote` module rather than doing it all on the main thread?

Here is the relevant code for anybody passing by: https://github.com/nylas/N1/blob/416f7e298d71a24515ed9d7dbe0ed354d9b5e287/src/nylas-env.coffee#L419

**Comment by Ciantic** [comment]
Not sure if this works, but in WinApi there is AnimateWindow for fading in and out: https://msdn.microsoft.com/en-us/library/windows/desktop/ms632669(v=vs.85).aspx

**Comment by bakkiraju** [comment]
Any recommended best practice to do fade in / fade out animations on OS X ?

**Comment by nickyhajal** [comment]
I'm also curious about this! I am currently using CSS but the graphics performance is noticeably laggy and with artifacts/tearing.

**Comment by bakkiraju** [comment]
@nickyhajal can you show me how you did it with css ?, I can only animate window contents inside not the window itself. thanks

**Comment by nickyhajal** [comment]
@bakkiraju Sorry for the confusion - to clarify, I'm doing what was discussed earlier in the thread: transparent window with a HTML/CSS "window".

When it covers a large portion of the screen, there are noticeable artifacts.

**Comment by bakkiraju** [comment]
@nickyhajal do u have a code snippet u can point me to ?, am new to web stuff and learning by example code, 

 PS: My electron windows are small and I want to see what you are able to do with CSS when it comes to animations. Thanks in advance

**Comment by bakkiraju** [comment]
@zcbenz trying to use electron and this kind of animation is a make it or break it for me, any recommendation or pointers on how I can fade out the BrowserWindow, using just the CSS on a transparent, frameless window doesn't seem to help, it's still fading only the contents inside BrowserWindow. Thanks

**Comment by iLeonidze** [comment]
I'm animating window using jQuery from renderer. You can modify & use this code to fade.

Code:
```var electron = require('electron').remote;
      var Chart = require('chart.js');
      const BrowserWindow = electron.BrowserWindow;
      var electronVibrancy = require('./electron-vibrancy');
      var path = require('path');
      const jQuery = require('jquery');

      // Set easigns
      jQuery.easing.jswing = jQuery.easing.swing;
      jQuery.extend(jQuery.easing,
          {
              def: 'easeOutQuad',
              swing: function (x, t, b, c, d) {
                  //alert($.easing.default);
                  return jQuery.easing[jQuery.easing.def](x, t, b, c, d);
              },
              easeInOutExpo: function (x, t, b, c, d) {
                  if (t==0) return b;
                  if (t==d) return b+c;
                  if ((t/=d/2) < 1) return c/2 * Math.pow(2, 10 * (t - 1)) + b;
                  return c/2 * (-Math.pow(2, -10 * --t) + 2) + b;
              }
          });

      function animateAppIn(){
          return animateWindow("in");
      }
      function animateAppOut(){
          return animateWindow("out");
      }

      function animateWindow(direction){
          var win = BrowserWindow.getAllWindows()[0];
          var x = win.getPosition()[0];
          var y = win.getPosition()[1];
          var from, to;
          if(direction=="in") {
              from = electron.screen.getPrimaryDisplay().workArea.width;
              to = electron.screen.getPrimaryDisplay().workArea.width - jQuery(window).width();
          }else{
              from = electron.screen.getPrimaryDisplay().workArea.width - jQuery(window).width();
              to = electron.screen.getPrimaryDisplay().workArea.width;
          }
          jQuery({ Counter: from }).animate({ Counter: to }, {
              duration: 500,
              easing: 'easeOutExpo',
              step: function () {
                  win.setPosition(Math.ceil(this.Counter),y);
              }
          });
      }
```

Result:
![1](https://cloud.githubusercontent.com/assets/3584432/23586040/2938f7a0-019e-11e7-926a-56d094029160.gif)

**Comment by jetlej** [comment]
@iLeonidze - I'm getting this error with your code: 

`jQuery.easing[this.easing] is not a function`

Did you have to include jquery-ui? And if so, how did you do that?

**Comment by doublenb** [comment]
@jetlej - after some tinkering, the answer is yes. I had to install:

`npm install jsdom jquery jquery-ui --save`

Code:
```
var jsdom = require('jsdom');
const { JSDOM } = jsdom;
const { window } = new JSDOM();
const { document } = (new JSDOM('<!DOCTYPE html><html><head></head><body></body></html>')).window;
global.document = document;

var $ = jQuery = require('jquery')(window);

require("jquery-ui");
require("jquery-ui/ui/effect.js");
```

**Comment by pjebs** [comment]
This is what I ended up doing inside **renderer:**

```go
		currentWindow :=js.Global.Get("npm").Get("remote").Call("getCurrentWindow")
...
		totalSteps := 20
		totalTime := 1000 * time.Millisecond
		c := time.Tick(totalTime / time.Duration(totalSteps))
		currentOpacity := currentWindow.Call("getOpacity").Float()
		go func() {
			for range c {
				currentOpacity = currentOpacity + (1.0 / float64(totalSteps))
				currentWindow.Call("setOpacity", currentOpacity)
				if currentOpacity > 1 {
					break
				}
			}
		}()
```

The BrowserWindow's `opacity` is initially set to `0.0` (transparent).

It works quite well surprisingly.

The above code is in Go (which I transpile to JS). I'm sure you can figure out the equivalent in JS.

**Comment by ahkohd** [comment]
I wrote this function to fade out
```ts
const  fadeWindowOut =  (
    _window: BrowserWindow,
    step: number = 0.1,
    fadeEveryXSeconds: number = 10
  ) =>  {
    let opacity = 1;
    const interval = setInterval(() => {
      if (opacity <= 0) window.clearInterval(interval);
      _window.setOpacity(opacity);
      opacity -= step;
    }, fadeEveryXSeconds);
    return interval;
  }
```
